src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,194 @@ internal class DiagnosticComputer
                 }
             }
 
-            var diagnosticsComputer = new DiagnosticComputer(document, project,
-                solutionChecksum, ideOptions, span, analysisKind, analyzerInfoCache, hostWorkspaceServices);
-            return diagnosticsComputer.GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+            // We perform prioritized execution of diagnostic computation requests based on the
+            // 'highPriority' boolean parameter.
+            //   - High priority requests forces cancellation of all the executing normal priority requests,
+            //     which are re-attempted once the high priority request completes.
+            //   - Normal priority requests wait for all the executing high priority requests to complete
+            //     before starting the compute.
+            //   - Canceled normal priority requests are re-attempted in the below loop.
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - High priority task forces cancellation of all the executing normal priority tasks
+                //    to minimize resource and CPU contention with normal priority tasks.
+                //  - Normal priority task waits for all the executing high priority tasks to complete.
+                if (highPriority)
+                {
+                    CancelNormalPriorityTasks(cancellationToken);
```

Comment:
>CancelNormalPriorityTasks(cancellationToken); [](http://example.com/codeflow?start=20&length=45)

Should the high priority task be added to s_highPriorityComputeTasks before canceling the normal priority tasks? Otherwise, couldn't the normal priority task loop continue and think there aren't any high-priority tasks?

Reply:
Thanks Todd, that is great point. We leave a small window where normal priority task can sneak in. This makes me believe we need a separate ordering of operations for high priority task. I’ll take up prior suggestion from Cyrus and break out the logic into two separate code paths for high priority and normal priority tasks.

Reply:
For my own understanding, I tried coding out Cyrus's suggestion and came up with something like 

https://github.com/ToddGrun/roslyn/commit/72479b55f484be106cfe7aadf8b8b7741886467b

I'm curious how closely this matches what you come up with.

Reply:
Yes.  This is very close to *exactly* what i was envisioning.  left comments on htat commit.

Reply:
I've gone ahead and created a pull request from my branch into this branch with these proposed changes.

https://github.com/mavasani/roslyn/pull/4

@mavasani -- Please take a look and merge if you think it makes things better. I haven't been able to actually run with the changes, so I'm not 100% certain I haven't regressed anything, but the Cyrus's idea seems promising.

Reply:
@ToddGrun - Thank you for taking out time to code up the proposal. However, I need to first apologize that my prior comment https://github.com/dotnet/roslyn/pull/67392#discussion_r1145166818 was not clear. Cyrus had multiple suggestions in the review, and I wasn't talking about taking up Cyrus' suggestion where we sprinkle wait for high priority execution within the normal pri execution code path, primarily because that doesn't address the core perf issue. I'll elaborate in the next paragraph. I was talking about Cyrus' suggestion to just have 2 separate code paths for normal and high pri execution, which should be a minor refactoring, something I hope to push to the branch soon.

Now, getting to the core issue regarding why the approach in your proposed PR doesn't address the core performance issue. 99% of the diagnostic computation cost inside `DiagnosticComputer.GetDiagnosticsAsync` comes from the below single invocation into the compiler layer: https://github.com/dotnet/roslyn/blob/efbc0838757cd5faedce55bb866ce1ad9332b4a7/src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs#L160-L161
Adding `await waitForPriorityExecutionAsync(cancellationToken).ConfigureAwait(false);` calls before or after that call has no meaningful perf impact. Even if we desired to make a public API change in `CompilationWithAnalyzers` to take this waitable Task as an input parameter, it won't solve our issue. For example, one of the large contributors of execution time is the core binding cost within the compiler layer, and we would never be able to sprinkle these wait calls into the core compiler. Fortunately enough, all these long running operations already support cancellation mid-way via the `CancellationToken` parameter. The approach chosen in the PR takes advantage of that and hooks up a custom CancellationTokenSource which can fire cancellation anytime while we are executing the above cited call into `CompilationWithAnalyzers.GetAnalysisResultAsync`.

Please let me know if the above makes sense or is still unclear. We can sync up offline if required. I am going to take some core refactoring pieces from your commit, while leaving the core logic of using a linked cancellation token in place. Hope that sounds reasonable.

Reply:
What doe GetAnalysisResultAsync do, and can it be made cooperative?

Note: something *still* unclear to me is why concurrent calls to get the diagnostics by one feature (tagging or lsp-pull-diags) would be so detrimental to the explicit call to get diagnostics for lightbulbs.  On a low-core machine i can see a problem.  But on a normal 4-8+ core machine, it seems like this should all run concurrently with no issue.  

The reason i ask is that concurrency is the heart of how our features work.  We have things running concurrently all the time, with the necessity that (as long as something isn't abusing the resources), latency will still be maintained.  That isn't holding up here, and that's concerning.

Reply:
> What doe GetAnalysisResultAsync do, and can it be made cooperative?

This is the core public API exposed from CompilationWithAnalyzers to do all the work for diagnostic computation - force complete the file to produce declared symbol events in the file, bind method bodies to create IOp trees, and make all the analyzer callbacks for computing diagnostics.

> can it be made cooperative?

It is already cooperative by taking a `cancellationToken` parameter so any caller can cancel this long running operation midway through the compute.

> Note: something still unclear to me is why concurrent calls to get the diagnostics by one feature (tagging or lsp-pull-diags) would be so detrimental to the explicit call to get diagnostics for lightbulbs.

> Note: something still unclear to me is why concurrent calls to get the diagnostics by one feature (tagging or lsp-pull-diags) would be so detrimental to the explicit call to get diagnostics for lightbulbs.

I wish I could assertively answer this, but all the performance measurements and traces show that simultaneous calls are slowing down the time to complete the lightbulb call. One possible theory is there are lot of shared objects/data structures between these computations. For example, core diagnostic analyzer instances are shared between these calls. Things such as project-level AnalyzerConfig providers to compute effective severity of a diagnostic ID for a given file (via parsing editorconfigs) would be shared. I suspect operating with these shared data structures, along with locking and synchronization logic for multiple simultaneous requests is going to add to the cost.

> The reason i ask is that concurrency is the heart of how our features work. We have things running concurrently all the time, with the necessity that (as long as something isn't abusing the resources), latency will still be maintained. That isn't holding up here, and that's concerning.

I guess this is true for multiple disparate computations that do not share objects/data structures that need synchronized access, locking, etc. For computations that do share these, one can definitely cause performance impact in completion of the other, and the compute time for one will definitely be lesser when there is no such contention.